kernel/typeops.ml

```diff
@@ -591,6 +617,15 @@ let rec execute env cstr =
     (* Primitive types *)
     | Int _ -> cstr, type_of_int env
     | Float _ -> cstr, type_of_float env
+    | Array(t,def) ->
+      let def', ty = execute env def in
+      let ta = type_of_array env in
+      let t' = Array.Smart.map (fun x ->
+        let x', xt = execute env x in
+        check_cast env x' xt DEFAULTcast ty;
+        x') t in
+      let cstr = if def'==def && t'==t then cstr else mkArray(t',def') in
+      cstr, mkApp(ta, [|ty|])
```

Comment:
Once we can put types bigger then Set in arrays this seems incorrect, ie 
~~~coq
Fail Check [| Set | nat |].
~~~
ie we need to somehow find a common supertype instead of assuming that the inferred type of the default is what we want.

We also lose subject reduction:
~~~coq
Definition u := Type.
Definition x : array u := [| | nat : u |].
Fail Definition y : array u := Eval lazy in x.
~~~
the eval removes the cast, so we get `[| | nat |] : array Set` and `array Set <> array u`.

We probably need to include the `ty` as a subterm of the array literal to avoid this.